diff --git a/curr/st.h b/other/st.h
index e4ce698..c32f755 100644
--- a/curr/st.h
+++ b/other/st.h
@@ -3,6 +3,10 @@
 #include <stdint.h>
 #include <sys/types.h>
 
+/* Arbitrary sizes */
+#define UTF_SIZ     4
+#define ESC_BUF_SIZ (128 * UTF_SIZ)
+
 /* macros */
 #define MIN(a, b)        ((a) < (b) ? (a) : (b))
 #define MAX(a, b)        ((a) < (b) ? (b) : (a))
@@ -60,7 +64,6 @@ typedef unsigned short ushort;
 
 typedef uint_least32_t Rune;
 
-#define Glyph Glyph_
 typedef struct {
 	Rune     u;    /* character code */
 	ushort   mode; /* attribute flags */
@@ -75,7 +78,6 @@ typedef union {
 	uint        ui;
 	float       f;
 	const void *v;
-	const char *s;
 } Arg;
 
 void die(const char *, ...);
@@ -92,7 +94,7 @@ void   tnew(int, int);
 void   tresize(int, int);
 void   tsetdirtattr(int);
 void   ttyhangup(void);
-int    ttynew(const char *, char *, const char *, char **);
+int    ttynew(char *, char *, char *, char **);
 size_t ttyread(void);
 void   ttyresize(int, int);
 void   ttywrite(const char *, size_t, int);
@@ -106,22 +108,20 @@ void  selextend(int, int, int, int);
 int   selected(int, int);
 char *getsel(void);
 
+size_t utf8decode(const char *, Rune *, size_t);
 size_t utf8encode(Rune, char *);
 
 void *xmalloc(size_t);
 void *xrealloc(void *, size_t);
-char *xstrdup(const char *);
+char *xstrdup(char *);
 
 /* config.h globals */
 extern char        *utmp;
-extern char        *scroll;
 extern char        *stty_args;
 extern char        *vtiden;
 extern wchar_t     *worddelimiters;
 extern int          allowaltscreen;
-extern int          allowwindowops;
 extern char        *termname;
 extern unsigned int tabspaces;
 extern unsigned int defaultfg;
 extern unsigned int defaultbg;
-extern unsigned int defaultcs;
